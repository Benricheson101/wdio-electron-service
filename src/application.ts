import fs from 'node:fs/promises';
import path from 'node:path';

import type { NormalizedReadResult } from 'read-package-up';

import log from './log.js';
import { APP_NAME_DETECTION_ERROR, BUILD_TOOL_DETECTION_ERROR, MULTIPLE_BUILD_TOOLS_ERROR } from './constants.js';
import type {
  AppBuildInfo,
  ElectronBuilderArch,
  ElectronBuilderConfig,
  ElectronForgeConfig,
  ForgeArch,
} from './types.js';
import { allOfficialArchsForPlatformAndVersion } from '@electron/packager';

const SupportedPlatform = {
  darwin: 'darwin',
  linux: 'linux',
  win32: 'win32',
};

/**
 * Determine the path to the Electron application binary
 * @param packageJsonPath path to the nearest package.json
 * @param appName name of the application
 * @param buildToolConfig configuration for the detected build tool
 * @param p   process object (used for testing purposes)
 * @returns   path to the Electron app binary
 */
export async function getBinaryPath(
  packageJsonPath: string,
  appBuildInfo: AppBuildInfo,
  electronVersion?: string,
  p = process,
) {
  if (!Object.values(SupportedPlatform).includes(p.platform)) {
    throw new Error(`Unsupported platform: ${p.platform}`);
  }

  let outDirs: string[];

  if (appBuildInfo.isForge) {
    // Forge case
    const archs = allOfficialArchsForPlatformAndVersion(
      p.platform as keyof typeof SupportedPlatform,
      electronVersion,
    ) as ForgeArch[];

    // Electron Forge always bundles into an `out` directory - see comment in getAppBuildInfo below
    outDirs = archs.map((arch) =>
      path.join(path.dirname(packageJsonPath), 'out', `${appBuildInfo.appName}-${p.platform}-${arch}`),
    );
  } else {
    // electron-builder case
    const archs: ElectronBuilderArch[] = ['arm64', 'armv7l', 'ia32', 'universal', 'x64'];
    const builderOutDirName = (appBuildInfo.config as ElectronBuilderConfig)?.directories?.output || 'dist';
    const builderOutDirMap = (arch: ElectronBuilderArch) => ({
      darwin: path.join(builderOutDirName, arch === 'x64' ? 'mac' : `mac-${p.arch}`),
      linux: path.join(builderOutDirName, 'linux-unpacked'),
      win32: path.join(builderOutDirName, 'win-unpacked'),
    });

    outDirs = archs.map((arch) =>
      path.join(path.dirname(packageJsonPath), builderOutDirMap(arch)[p.platform as keyof typeof SupportedPlatform]),
    );
  }

  const binaryPathMap = {
    darwin: path.join(`${appBuildInfo.appName}.app`, 'Contents', 'MacOS', appBuildInfo.appName),
    linux: appBuildInfo.appName,
    win32: `${appBuildInfo.appName}.exe`,
  };
  const electronBinaryPath = binaryPathMap[p.platform as keyof typeof SupportedPlatform];

  const binaryPaths = outDirs.map((outDir) => path.join(outDir, electronBinaryPath));

  // for each path, check if it exists and is executable
  const executableBinaryPaths = binaryPaths.filter(async (binaryPath) => {
    try {
      await fs.access(binaryPath, fs.constants.X_OK);
      return true;
    } catch (e) {
      log.debug(e);
      return false;
    }
  });

  // no binary case
  if (executableBinaryPaths.length === 0) {
    throw new Error(`No executable binary found, checked [${binaryPaths.join(', ')}]`);
  }

  // multiple binaries case
  if (executableBinaryPaths.length > 1) {
    log.debug(`Detected multiple app binaries, using the first one: ${executableBinaryPaths[0]}`);
  }

  return executableBinaryPaths[0];
}

/**
 * Determine build information about the Electron application
 * @param pkg path to the nearest package.json
 * @returns   promise resolving to the app build information
 */
export async function getAppBuildInfo(pkg: NormalizedReadResult): Promise<AppBuildInfo> {
  const forgeDependencyDetected = Object.keys(pkg.packageJson.devDependencies || {}).includes('@electron-forge/cli');
  const builderDependencyDetected = Object.keys(pkg.packageJson.devDependencies || {}).includes('electron-builder');

  // Forge configuration is not currently used to determine the Electron app binary path
  // - when custom output directories are supported in Forge we can use this config value for path determination
  // - see https://github.com/electron/forge/pull/2714
  const forgePackageJsonConfig = pkg.packageJson.config?.forge;
  const forgeCustomConfigFile = typeof forgePackageJsonConfig === 'string';
  const forgeConfigPath = forgeCustomConfigFile ? forgePackageJsonConfig : 'forge.config.js';
  const rootDir = path.dirname(pkg.path);
  let forgeConfig = forgePackageJsonConfig as ElectronForgeConfig;
  let builderConfig: ElectronBuilderConfig = pkg.packageJson.build;

  if (!forgePackageJsonConfig || forgeCustomConfigFile) {
    // if no config or a linked file, attempt to read Forge JS-based config
    try {
      log.debug(`Reading Forge config file: ${forgeConfigPath}...`);
      forgeConfig = ((await import(path.join(rootDir, forgeConfigPath))) as { default: ElectronForgeConfig }).default;
    } catch (e) {
      log.debug(e);
    }
  }

  const isForge = Boolean(forgeConfig || forgeDependencyDetected);

  if (!isForge) {
    // if no Forge config or dependency, attempt to read `electron-builder.json`
    try {
      log.debug('Forge not detected, reading `electron-builder.json`...');
      const data = await fs.readFile(path.join(rootDir, 'electron-builder.json'), 'utf-8');
      builderConfig = JSON.parse(data);
    } catch (e) {
      log.debug(e);
    }
  }

  const isBuilder = Boolean(builderConfig || builderDependencyDetected);

  if (isForge && isBuilder) {
    throw new Error(MULTIPLE_BUILD_TOOLS_ERROR);
  }
  if (!isForge && !isBuilder) {
    throw new Error(BUILD_TOOL_DETECTION_ERROR);
  }

  const config = isForge ? forgeConfig : builderConfig;
  log.debug(`${isForge ? 'Forge' : 'Builder'} configuration detected: ${config}`);

  const appName: string =
    pkg.packageJson.productName ||
    (isBuilder && (config as ElectronBuilderConfig)?.productName) ||
    (isForge && (config as ElectronForgeConfig)?.packagerConfig?.name) ||
    pkg.packageJson.name;

  if (!appName) {
    throw new Error(APP_NAME_DETECTION_ERROR);
  }

  return {
    appName,
    config,
    isForge,
    isBuilder,
  };
}
