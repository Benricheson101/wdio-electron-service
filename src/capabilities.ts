import type { Capabilities } from '@wdio/types';

import type { ElectronServiceOptions } from './types';

export function getChromeOptions(options: ElectronServiceOptions, cap: Capabilities.Capabilities) {
  const existingOptions = cap['goog:chromeOptions'] || {};
  return {
    binary: options.appBinaryPath,
    windowTypes: ['app', 'webview'],
    ...existingOptions,
    args: [...(existingOptions.args || []), ...(options.appArgs || [])],
  };
}

export function getChromedriverOptions(cap: Capabilities.Capabilities) {
  const existingOptions = cap['wdio:chromedriverOptions'] || {};
  return existingOptions;
}

const isElectron = (cap: unknown) =>
  (cap as Capabilities.DesiredCapabilities)?.browserName?.toLowerCase() === 'electron';

/**
 * get capability independent of which type of capabilities is set
 */
export function getElectronCapabilities(caps: Capabilities.RemoteCapability) {
  /**
   * standard capabilities, e.g.:
   * ```
   * {
   *   browserName: 'chrome'
   * }
   * ```
   */
  const standardCaps = caps as Capabilities.Capabilities;
  if (typeof standardCaps.browserName === 'string' && isElectron(standardCaps)) {
    return [caps as Capabilities.Capabilities];
  }
  /**
   * W3C specific capabilities, e.g.:
   * ```
   * {
   *   alwaysMatch: {
   *     browserName: 'chrome'
   *   }
   * }
   * ```
   */
  const w3cCaps = (caps as Capabilities.W3CCapabilities).alwaysMatch;
  if (w3cCaps && typeof w3cCaps.browserName === 'string' && isElectron(w3cCaps)) {
    return [w3cCaps];
  }
  /**
   * multiremote capabilities, e.g.:
   * ```
   * {
   *   instanceA: {
   *     capabilities: {
   *        browserName: 'chrome'
   *     }
   *   },
   *   instanceB: {
   *     capabilities: {
   *        browserName: 'chrome'
   *     }
   *   }
   * }
   * ```
   */
  return Object.values(caps as Capabilities.MultiRemoteCapabilities)
    .map(
      (options) =>
        (options.capabilities as Capabilities.W3CCapabilities)?.alwaysMatch ||
        (options.capabilities as Capabilities.Capabilities),
    )
    .filter((caps) => isElectron(caps));
}
