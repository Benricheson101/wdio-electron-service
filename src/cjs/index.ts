import type { Capabilities, Options, Services } from '@wdio/types';
import type { default as LauncherInstance, ElectronLauncherServiceOpts } from '../launcher.js';
import type { default as ServiceInstance } from '../service.js';

exports.default = class CJSElectronService {
  private instance?: ServiceInstance;

  constructor(options: Services.ServiceOption) {
    (async () => {
      const { default: ElectronService } = await import('../service.js');
      this.instance = new ElectronService(options);
    })();
  }

  async beforeSession(config: Omit<Options.Testrunner, 'capabilities'>, capabilities: Capabilities.Capabilities) {
    const instance = await this.instance;
    return instance?.beforeSession(config, capabilities);
  }

  async before(capabilities: Capabilities.Capabilities, specs: string[], browser: WebdriverIO.Browser) {
    const instance = await this.instance;
    return instance?.before(capabilities, specs, browser);
  }
};

exports.launcher = class CJSElectronServiceLauncher {
  private instance?: LauncherInstance;

  constructor(
    options: ElectronLauncherServiceOpts,
    capabilities: Capabilities.Capabilities,
    config: Options.Testrunner,
  ) {
    (async () => {
      const { default: ElectronServiceLauncher } = await import('../launcher.js');
      this.instance = new ElectronServiceLauncher(options, capabilities, config);
    })();
  }

  async onPrepare() {
    const instance = await this.instance;
    return instance?.onPrepare();
  }

  async onComplete() {
    const instance = await this.instance;
    return instance?.onComplete();
  }
};
