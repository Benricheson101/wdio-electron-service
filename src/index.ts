import { browser as wdioBrowser } from '@wdio/globals';
import ElectronWorkerService from './service.js';

export default ElectronWorkerService;
export interface BrowserExtension {
  electron: {
    api: (...arg: unknown[]) => Promise<unknown>;
    app: (funcName: string, ...arg: unknown[]) => Promise<unknown>;
    browserWindow: (funcName: string, ...arg: unknown[]) => Promise<unknown>;
    dialog: (funcName: string, ...arg: unknown[]) => Promise<unknown>;
    mainProcess: (funcName: string, ...arg: unknown[]) => Promise<unknown>;
    mock: (apiName: string, funcName: string, mockReturnValue: unknown) => Promise<unknown> | unknown;
  };
}

declare global {
  namespace WebdriverIO {
    interface Browser extends BrowserExtension {}
    interface MultiRemoteBrowser extends BrowserExtension {}
  }
}

export const browser: WebdriverIO.Browser = wdioBrowser;
