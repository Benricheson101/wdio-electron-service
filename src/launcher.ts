import { join } from 'path';
import { Capabilities } from '@wdio/types';
import { launcher as ChromedriverServiceLauncher, ChromedriverServiceOptions } from 'wdio-chromedriver-service';

type WdioConfig = {
  [key: string]: unknown;
};

type ElectronLauncherServiceOpts = {
  chromedriver?: ChromedriverServiceOptions;
};

function createChromedriverServiceOptions(
  options: ElectronLauncherServiceOpts,
  resolver: NodeJS.RequireResolve,
): ChromedriverServiceOptions {
  const { chromedriver = {} } = options;
  const chromedriverServiceOptions = { ...chromedriver };

  process.env.WDIO_ELECTRON = 'true';

  if (!chromedriverServiceOptions.chromedriverCustomPath) {
    try {
      const electronChromeDriverPath = resolver('electron-chromedriver/chromedriver');
      chromedriverServiceOptions.chromedriverCustomPath = electronChromeDriverPath;
    } catch (e) {
      throw new Error(
        'electron-chromedriver was not found. You need to install it or provide a binary via the chromedriver.chromedriverCustomPath option.',
      );
    }
  }

  return chromedriverServiceOptions;
}

export default class ChromeDriverLauncher extends ChromedriverServiceLauncher {
  constructor(
    options: ElectronLauncherServiceOpts,
    capabilities: Capabilities.Capabilities,
    config: WdioConfig,
    resolver = require.resolve,
  ) {
    const isWin = process.platform === 'win32';
    const chromedriverServiceOptions = createChromedriverServiceOptions(options, resolver);

    if (isWin) {
      process.env.WDIO_ELECTRON_NODE_PATH = process.execPath;
      process.env.WDIO_ELECTRON_CHROMEDRIVER_PATH = chromedriverServiceOptions.chromedriverCustomPath;
      chromedriverServiceOptions.chromedriverCustomPath = join(__dirname, '..', 'bin', 'chrome-driver.bat');
    }

    super(chromedriverServiceOptions, capabilities, config);
  }
}
