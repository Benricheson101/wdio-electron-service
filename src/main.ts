import electron, { App, app, BrowserWindow, dialog, Dialog, ipcMain } from 'electron';

type AppFunction = (this: App, ...args: unknown[]) => unknown;
type MainProcessFunction = (this: NodeJS.Process, ...args: unknown[]) => unknown;
type BrowserWindowFunction = (this: BrowserWindow, ...args: unknown[]) => unknown;

let showMessageBoxResponse = { response: 1, checkboxChecked: true };

dialog.showMessageBox = ((
  browserWindow: BrowserWindow,
  { message }: Electron.MessageBoxOptions,
): Promise<Electron.MessageBoxReturnValue> => Promise.resolve(showMessageBoxResponse)) as Dialog['showMessageBox'];

ipcMain.handle('wdio-electron.mainProcess', (_event, funcName: string, ...args: unknown[]) => {
  const processProp = process[funcName as keyof NodeJS.Process];
  if (typeof processProp === 'function') {
    return (processProp as MainProcessFunction).apply(process, args);
  }
  return processProp;
});

ipcMain.handle('wdio-electron.app', (_event, funcName: string, ...args: unknown[]) => {
  const appProp = app[funcName as keyof App];
  if (typeof appProp === 'function') {
    return (appProp as AppFunction).apply(app, args);
  }
  return appProp;
});

ipcMain.handle('wdio-electron.browserWindow', (event, funcName: string, ...args: unknown[]) => {
  const browserWindow = BrowserWindow.fromWebContents(event.sender) as BrowserWindow;
  const browserWindowProp = browserWindow[funcName as keyof BrowserWindow];
  if (typeof browserWindowProp === 'function') {
    return (browserWindowProp as BrowserWindowFunction).apply(browserWindow, args);
  }
  return browserWindowProp;
});

ipcMain.handle('wdio-electron.dialog', (_event, funcName: string, ...args: unknown[]) => {
  const dialogProp = dialog[funcName as keyof Dialog];
  if (typeof dialogProp === 'function') {
    return (dialogProp as AppFunction).apply(app, args);
  }
  return dialogProp;
});

ipcMain.handle('wdio-electron.mainProcess', (_event, funcName: string, ...args: unknown[]) => {
  const processProp = process[funcName as keyof NodeJS.Process];
  if (typeof processProp === 'function') {
    return (processProp as MainProcessFunction).apply(process, args);
  }
  return processProp;
});

ipcMain.handle('wdio-electron.mock', (_event, apiName: string, funcName: string, mockReturnValue: unknown) => {
  const electronApi = electron[apiName as keyof typeof electron];
  const electronApiFunc = electronApi[funcName as keyof typeof electronApi];
  if (typeof electronApiFunc !== 'function') {
    throw new Error(`Unable to find function ${funcName} on ${apiName} module.`);
  }

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  electron[apiName][funcName] = funcName.endsWith('Sync')
    ? () => mockReturnValue
    : () => Promise.resolve(mockReturnValue);

  return {
    apiName,
    funcName,
    mockReturnValue,
  };
});
