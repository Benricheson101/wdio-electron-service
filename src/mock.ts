import { fn as vitestFn, type Mock } from '@vitest/spy';
import type { ElectronInterface, ElectronType } from './types.js';
import log from './log.js';

type BasicFn = (...args: unknown[]) => unknown;
type BasicAsyncFn = (...args: unknown[]) => Promise<unknown>;
type AbstractFn = BasicFn | BasicAsyncFn;
type ElectronApiFn = ElectronType[ElectronInterface][keyof ElectronType[ElectronInterface]];

type Omitted =
  | 'mockImplementation'
  | 'mockImplementationOnce'
  | 'mockReturnValue'
  | 'mockReturnValueOnce'
  | 'mockResolvedValue'
  | 'mockResolvedValueOnce'
  | 'mockRejectedValue'
  | 'mockRejectedValueOnce'
  | 'withImplementation';

interface AsyncMockInstance extends Omit<Mock, Omitted> {
  mockImplementation(fn: AbstractFn): Promise<AsyncMock>;
  mockImplementationOnce(fn: AbstractFn): Promise<AsyncMock>;
  mockReturnValue(obj: unknown): Promise<AsyncMock>;
  mockReturnValueOnce(obj: unknown): Promise<AsyncMock>;
  mockClear(): any;
  mockReset(): any;
  mockRestore(): Promise<AsyncMock>;
  update(): Promise<AsyncMock>;
  updating: boolean;
}

export interface AsyncMock<TArgs extends any[] = any, TReturns = any> extends AsyncMockInstance {
  new (...args: TArgs): TReturns;
  (...args: TArgs): TReturns;
}

async function setElectronMock(
  apiName: string,
  funcName: string,
  implementationFn: AbstractFn = () => undefined,
  returnValue?: unknown,
): Promise<void> {
  await browser.electron.execute(
    (electron, apiName, funcName, mockImplementation, mockReturnValue) => {
      const electronApi = electron[apiName as keyof typeof electron];
      const mockFn = globalThis.fn(eval(mockImplementation));

      if (mockReturnValue !== undefined) {
        mockFn.mockReturnValue(mockReturnValue);
      }

      // replace target API with mock
      electronApi[funcName as keyof typeof electronApi] = mockFn as ElectronApiFn;
    },
    apiName,
    funcName,
    implementationFn.toString(),
    returnValue,
  );
}

export function createMock(apiName: string, funcName: string) {
  const mock = vitestFn() as unknown as AsyncMock;
  const originalMockImplementation = mock.mockImplementation;
  const originalMockClear = mock.mockClear;
  const originalMockReset = mock.mockReset;

  mock.updating = false;
  mock.mockName(`electron.${apiName}.${funcName}`);

  mock.mockImplementation = async (fn: AbstractFn) => {
    await setElectronMock(apiName, funcName, fn);
    originalMockImplementation(fn);
    return mock;
  };

  mock.mockImplementationOnce = async (fn: AbstractFn) => {
    const implementationFn = (...args: unknown[]) => {
      mock.mockRestore();
      return fn.apply(fn, args);
    };
    await setElectronMock(apiName, funcName, implementationFn);
    originalMockImplementation(implementationFn);
    return mock;
  };

  mock.mockReturnValue = async (obj: unknown) => {
    await setElectronMock(apiName, funcName, undefined, obj);
    return mock;
  };

  mock.mockReturnValueOnce = async (obj: unknown) => {
    const implementationFn = () => mock.mockRestore();
    await setElectronMock(apiName, funcName, implementationFn, obj);
    originalMockImplementation(implementationFn);
    return mock;
  };

  mock.update = async () => {
    mock.updating = true;
    const calls = await browser.electron.execute(
      (electron, apiName, funcName) => {
        const mockObj = electron[apiName as keyof typeof electron][
          funcName as keyof ElectronType[ElectronInterface]
        ] as AsyncMock;
        return mockObj.mock?.calls || [];
      },
      apiName,
      funcName,
    );

    // re-apply calls from the electron main process mock to the outer one
    if (mock.mock.calls.length < calls.length) {
      calls.forEach((call: unknown[], index: number) => {
        if (!mock.mock.calls[index]) {
          mock?.apply(mock, call);
        }
      });
    }
    mock.updating = false;

    return mock;
  };

  mock.mockClear = async () => {
    await browser.electron.execute(
      (electron, apiName, funcName) => {
        (
          electron[apiName as keyof typeof electron][funcName as keyof ElectronType[ElectronInterface]] as Mock
        ).mockClear();
      },
      apiName,
      funcName,
    );
    originalMockClear();
    return mock;
  };

  mock.mockReset = async () => {
    await browser.electron.execute(
      (electron, apiName, funcName) => {
        (electron[apiName as keyof typeof electron][funcName as keyof ElectronType[ElectronInterface]] as Mock)
          .mockReset()
          .mockClear();
      },
      apiName,
      funcName,
    );
    originalMockReset();
    originalMockClear();
    return mock;
  };

  mock.mockRestore = async () => {
    // restores inner implementation to the original function
    await browser.electron.execute(
      async (electron, apiName, funcName) => {
        const electronApi = electron[apiName as keyof typeof electron];
        const originalApi = globalThis.originalApi as Record<ElectronInterface, ElectronType[ElectronInterface]>;
        const originalApiMethod = originalApi[apiName as keyof typeof originalApi][
          funcName as keyof ElectronType[ElectronInterface]
        ] as ElectronApiFn;

        if (originalApiMethod) {
          electronApi[funcName as keyof typeof electronApi] = (originalApiMethod as () => unknown).bind({}) as never;
        }
      },
      apiName,
      funcName,
    );

    return mock;
  };

  return mock;
}
