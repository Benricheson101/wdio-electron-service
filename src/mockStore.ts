import type { AsyncMock } from './mock.js';

export class ElectronServiceMockStore {
  #mockFns: Map<string, AsyncMock>;

  constructor() {
    this.#mockFns = new Map<string, AsyncMock>();
  }

  setMock(mock: AsyncMock): AsyncMock {
    this.#mockFns.set(mock.getMockName(), mock);
    return mock;
  }

  getMock(mockId: string) {
    const mock = this.#mockFns.get(mockId);
    if (!mock) {
      throw new Error(`No mock registered for "${mockId}"`);
    }

    return mock;
  }

  getMocks() {
    return Array.from(this.#mockFns.entries());
  }

  removeMock(mockId: string) {
    this.#mockFns.delete(mockId);
  }
}

const mockStore = new ElectronServiceMockStore();

export default mockStore;
