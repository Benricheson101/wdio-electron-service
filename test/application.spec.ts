import fs from 'node:fs/promises';
import path from 'node:path';

import { describe, it, expect } from 'vitest';

import { getBinaryPath, getAppBuildInfo } from '../src/application';

describe('getBinaryPath', () => {
  const pkgJSONPath = '/foo/bar/package.json';
  const winProcess = { arch: 'x64', platform: 'win32' } as NodeJS.Process;
  const macProcess = { arch: 'arm64', platform: 'darwin' } as NodeJS.Process;
  const linuxProcess = { arch: 'arm', platform: 'linux' } as NodeJS.Process;
  const unsupportedPlatformProcess = { platform: 'aix', arch: 'mips' } as NodeJS.Process;

  it('should throw an error when provided with an unsupported platform', async () =>
    expect(() =>
      getBinaryPath(
        pkgJSONPath,
        {
          appName: 'my-app',
          config: 'path/to/forge-config.js',
          isForge: true,
          isBuilder: false,
        },
        unsupportedPlatformProcess,
      ),
    ).rejects.toThrow('Unsupported platform: aix'));

  it('should return the expected app path for an Electron Forge setup on Windows', async () =>
    expect(
      await getBinaryPath(
        pkgJSONPath,
        {
          appName: 'my-app',
          config: 'path/to/forge-config.js',
          isForge: true,
          isBuilder: false,
        },
        winProcess,
      ),
    ).toBe(path.join('/foo', 'bar', 'out', 'my-app-win32-x64', 'my-app.exe')));

  it('should return the expected app path for an Electron Forge setup on Mac', async () =>
    expect(
      await getBinaryPath(
        pkgJSONPath,
        {
          appName: 'my-app',
          config: 'path/to/forge-config.js',
          isForge: true,
          isBuilder: false,
        },
        macProcess,
      ),
    ).toBe(path.join('/foo', 'bar', 'out', 'my-app-darwin-arm64', 'my-app.app', 'Contents', 'MacOS', 'my-app')));

  it('should return the expected app path for an Electron Forge setup on Linux', async () =>
    expect(
      await getBinaryPath(
        pkgJSONPath,
        {
          appName: 'my-app',
          config: 'path/to/forge-config.js',
          isForge: true,
          isBuilder: false,
        },
        linuxProcess,
      ),
    ).toBe(path.join('/foo', 'bar', 'out', 'my-app-linux-arm', 'my-app')));

  it('should return the expected app path for an electron-builder setup with custom output directory', async () =>
    expect(
      await getBinaryPath(
        pkgJSONPath,
        {
          appName: 'my-app',
          config: { productName: 'my-app', directories: { output: 'custom-outdir' } },
          isForge: false,
          isBuilder: true,
        },
        winProcess,
      ),
    ).toBe(path.join('/foo', 'bar', 'custom-outdir', 'win-unpacked', 'my-app.exe')));

  it('should return the expected app path for an electron-builder setup on Windows', async () =>
    expect(
      await getBinaryPath(
        pkgJSONPath,
        {
          appName: 'my-app',
          config: { productName: 'my-app' },
          isForge: false,
          isBuilder: true,
        },
        winProcess,
      ),
    ).toBe(path.join('/foo', 'bar', 'dist', 'win-unpacked', 'my-app.exe')));

  it('should return the expected app path for an electron-builder setup on Mac', async () =>
    expect(
      await getBinaryPath(
        pkgJSONPath,
        {
          appName: 'my-app',
          config: { productName: 'my-app' },
          isForge: false,
          isBuilder: true,
        },
        macProcess,
      ),
    ).toBe(path.join('/foo', 'bar', 'dist', 'mac-arm64', 'my-app.app', 'Contents', 'MacOS', 'my-app')));

  it('should return the expected app path for an electron-builder setup on Linux', async () =>
    expect(
      await getBinaryPath(
        pkgJSONPath,
        {
          appName: 'my-app',
          config: { productName: 'my-app' },
          isForge: false,
          isBuilder: true,
        },
        linuxProcess,
      ),
    ).toBe(path.join('/foo', 'bar', 'dist', 'linux-unpacked', 'my-app')));
});

describe('getBuildToolConfig', () => {
  it('should throw an error when no build tools are found', async () => {
    const packageJsonPath = path.join(__dirname, 'fixtures', 'no-build-tool', 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
    await expect(() =>
      getAppBuildInfo({
        packageJson,
        path: packageJsonPath,
      }),
    ).rejects.toThrow(
      'No build tool was detected, if the application is compiled at a different location, please specify the `appBinaryPath` option in your capabilities.',
    );
  });

  it('should throw an error when configuration for multiple build tools are found', async () => {
    const packageJsonPath = path.join(__dirname, 'fixtures', 'multiple-build-tools-config', 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
    await expect(() =>
      getAppBuildInfo({
        packageJson,
        path: packageJsonPath,
      }),
    ).rejects.toThrow(
      'Multiple build tools were detected, please remove configuration and dependencies for tools which are not being used to build your application.',
    );
  });

  it('should throw an error when dependencies for multiple build tools are found', async () => {
    const packageJsonPath = path.join(__dirname, 'fixtures', 'multiple-build-tools-dependencies', 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
    await expect(() =>
      getAppBuildInfo({
        packageJson,
        path: packageJsonPath,
      }),
    ).rejects.toThrow(
      'Multiple build tools were detected, please remove configuration and dependencies for tools which are not being used to build your application.',
    );
  });

  it('should throw an error when configuration for electron-builder is found alongside an Electron Forge dependency', async () => {
    const packageJsonPath = path.join(__dirname, 'fixtures', 'multiple-build-tools-wrong-config-1', 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
    await expect(() =>
      getAppBuildInfo({
        packageJson,
        path: packageJsonPath,
      }),
    ).rejects.toThrow(
      'Multiple build tools were detected, please remove configuration and dependencies for tools which are not being used to build your application.',
    );
  });

  it('should throw an error when configuration for Electron Forge is found alongside an electron-builder dependency', async () => {
    const packageJsonPath = path.join(__dirname, 'fixtures', 'multiple-build-tools-wrong-config-2', 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
    await expect(() =>
      getAppBuildInfo({
        packageJson,
        path: packageJsonPath,
      }),
    ).rejects.toThrow(
      'Multiple build tools were detected, please remove configuration and dependencies for tools which are not being used to build your application.',
    );
  });

  it('should throw an error when the app name is unable to be determined', async () => {
    const packageJsonPath = path.join(__dirname, 'fixtures', 'forge-dependency-inline-config', 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
    delete packageJson.name;
    delete packageJson.config.forge.packagerConfig;
    await expect(() =>
      getAppBuildInfo({
        packageJson,
        path: packageJsonPath,
      }),
    ).rejects.toThrow(
      'No application name was detected, please set name / productName in your package.json or build tool configuration.',
    );
  });

  it('should return the expected config for a Forge dependency with inline config', async () => {
    const packageJsonPath = path.join(__dirname, 'fixtures', 'forge-dependency-inline-config', 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
    await expect(
      await getAppBuildInfo({
        packageJson,
        path: packageJsonPath,
      }),
    ).toStrictEqual({
      appName: 'forge-dependency-inline-config',
      config: { packagerConfig: { name: 'forge-dependency-inline-config' } },
      isBuilder: false,
      isForge: true,
    });
  });

  it('should return the expected config for a Forge dependency with JS config', async () => {
    const packageJsonPath = path.join(__dirname, 'fixtures', 'forge-dependency-js-config', 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
    await expect(
      await getAppBuildInfo({
        packageJson,
        path: packageJsonPath,
      }),
    ).toStrictEqual({
      appName: 'forge-dependency-js-config',
      config: { packagerConfig: { name: 'forge-dependency-js-config' } },
      isBuilder: false,
      isForge: true,
    });
  });

  it('should return the expected config for a Forge dependency with linked JS config', async () => {
    const packageJsonPath = path.join(__dirname, 'fixtures', 'forge-dependency-linked-js-config', 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
    await expect(
      await getAppBuildInfo({
        packageJson,
        path: packageJsonPath,
      }),
    ).toStrictEqual({
      appName: 'forge-dependency-linked-js-config',
      config: { packagerConfig: { name: 'forge-dependency-linked-js-config' } },
      isBuilder: false,
      isForge: true,
    });
  });

  it('should return the expected config for an electron-builder dependency with inline config', async () => {
    const packageJsonPath = path.join(__dirname, 'fixtures', 'builder-dependency-inline-config', 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
    await expect(
      await getAppBuildInfo({
        packageJson,
        path: packageJsonPath,
      }),
    ).toStrictEqual({
      appName: 'builder-dependency-inline-config',
      config: { productName: 'builder-dependency-inline-config' },
      isBuilder: true,
      isForge: false,
    });
  });

  it('should return the expected config for an electron-builder dependency with JSON config', async () => {
    const packageJsonPath = path.join(__dirname, 'fixtures', 'builder-dependency-json-config', 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
    await expect(
      await getAppBuildInfo({
        packageJson,
        path: packageJsonPath,
      }),
    ).toStrictEqual({
      appName: 'builder-dependency-json-config',
      config: { productName: 'builder-dependency-json-config' },
      isBuilder: true,
      isForge: false,
    });
  });

  it('should return the expected config when there is no app name in the build tool config', async () => {
    const packageJsonPath = path.join(__dirname, 'fixtures', 'no-app-name-in-build-tool-config', 'package.json');
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
    await expect(
      await getAppBuildInfo({
        packageJson,
        path: packageJsonPath,
      }),
    ).toStrictEqual({
      appName: 'fixture-no-app-name-in-build-tool-config',
      config: { appId: 'no-app-name-in-build-tool-config' },
      isBuilder: true,
      isForge: false,
    });
  });
});
