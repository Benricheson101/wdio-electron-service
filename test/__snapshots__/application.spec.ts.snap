// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getBinaryPath > should throw an error when no binary is found for a Builder setup on MacOS 1`] = `
[Error: No executable binary found, checked: 
/foo/bar/dist/mac-arm64/my-app.app/Contents/MacOS/my-app, 
/foo/bar/dist/mac-armv7l/my-app.app/Contents/MacOS/my-app, 
/foo/bar/dist/mac-ia32/my-app.app/Contents/MacOS/my-app, 
/foo/bar/dist/mac-universal/my-app.app/Contents/MacOS/my-app, 
/foo/bar/dist/mac/my-app.app/Contents/MacOS/my-app]
`;

exports[`getBinaryPath > should throw an error when no binary is found for a Forge setup 1`] = `
[Error: No executable binary found, checked: 
/foo/bar/out/my-app-win32-ia32/my-app.exe, 
/foo/bar/out/my-app-win32-x64/my-app.exe, 
/foo/bar/out/my-app-win32-arm64/my-app.exe]
`;
